File: 00_IO.R

# Save/Load R object .rds file
saveRDS(data, paste0("data.rds"))
data <- readRDS(paste0("data.rds"))

# Save/Load R object binary workspace (.RData)
save(data, file = paste0("data.RData"))
load(paste0("data.RData"))

# Write/Read table TSV
write.table(data, paste0("data.tsv"), sep = "\t", quote=F, row.names = F)
data <- read.table(paste0("data.tsv"), sep = "\t", header = TRUE)

# Write/Read table CSV
write.table(data, paste0("data.csv"), sep = ",", quote=F, row.names = F)
data <- read.table(paste0("data.csv"), sep = ",", header = TRUE)

# Write/Read Excel file (openxlsx package)
library(openxlsx)
write.xlsx(data, file = paste0("data.xlsx"))
data <- read.xlsx(paste0("data.xlsx"), sheet = 1)

# Write/Read JSON file (jsonlite package)
library(jsonlite)
write_json(data, path = paste0("data.json"))
data <- read_json(path = paste0("data.json"))

# Read Delim
data <- read.delim("data.txt", header = TRUE, sep = "\t", stringsAsFactors = FALSE, na.strings = "NA")

# Save plot to PDF file
p<-ggplot(data, aes(x=x, y=y)) +
  geom_point(alpha = 1.2,size=3)+
  
pdf(paste0("data.pdf"), width=3, height=5)
print(p)
dev.off()



# Arrange plots in a grid
library(gridExtra)
grid <- grid.arrange(p1, p2, p3, p4, ncol = 2)

# Save the grid as a PDF
pdf("grid_plot.pdf", width = 10, height = 8)
grid.draw(grid)
dev.off()

File: Python/00_IO.py
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import json
import openpyxl

# Write/Read table TSV
data.to_csv('data.tsv', sep='\t', index=False, quoting=3)  # quoting=3 prevents quotes around strings
data = pd.read_csv('data.tsv', sep='\t')

# Write/Read table CSV
data.to_csv('data.csv', index=False, quoting=3)
data = pd.read_csv('data.csv')

# Write/Read Excel file
data.to_excel('data.xlsx', index=False)
data = pd.read_excel('data.xlsx')

# Write/Read JSON file
data.to_json('data.json', orient='records', lines=True)
with open('data.json') as f:
    data = pd.read_json(f)

# Read Delim (txt file with tab-separated values)
data = pd.read_csv('data.txt', sep='\t', header=0, na_values="NA")

# Save plot to PDF file
# Example plot
plt.figure(figsize=(6,4))
sns.scatterplot(data=data, x='x', y='y', alpha=1.2, size=3)

plt.savefig("data.pdf", format="pdf")

# Arrange plots in a grid
from matplotlib import gridspec

fig = plt.figure(figsize=(10, 8))
gs = gridspec.GridSpec(2, 2, figure=fig)

# Example plots
p1 = fig.add_subplot(gs[0, 0])
sns.scatterplot(data=data, x='x', y='y', ax=p1)

p2 = fig.add_subplot(gs[0, 1])
sns.scatterplot(data=data, x='x', y='y', ax=p2)

p3 = fig.add_subplot(gs[1, 0])
sns.scatterplot(data=data, x='x', y='y', ax=p3)

p4 = fig.add_subplot(gs[1, 1])
sns.scatterplot(data=data, x='x', y='y', ax=p4)

# Save the grid as a PDF
plt.savefig("grid_plot.pdf", format="pdf")
